### PR TITLE
Restructure and add functionality to PCLPointCloud2 filters Par…

### PR DESCRIPTION
**Pull request by mvieth**
- replace two applyFilter(indices) functions
- previously, the functions did not consider removed_indices_ and extract_removed_indices_
- applyFilter(indices) in RandomSample previously also did not consider negative_
- adapted code from the templated filters (hpp files) was used
- follow-up to pr #3483, see also issue #3471

### CODE REVIEW

**Comment by mvieth** [comment]
The check in RandomSample fails because it is tested whether the two filter functions (the one that takes a cloud and the one that takes indices) do the same thing, which (with this commit) they don't. My plan was to create another PR later with [this commit](https://github.com/mvieth/pcl/commit/bc4f1fbcf083806da4794f50be145927cc2bfa16) which makes the `applyFilter(cloud)` function use the `applyFilter(indices)` function.

File: filters/src/extract_indices.cpp [hunk 1]
```diff
@@ -146,39 +146,52 @@ pcl::ExtractIndices<pcl::PCLPointCloud2>::applyFilter (PCLPointCloud2 &output)
...
+      for (int fii = 0; fii < static_cast<int> (full_indices.size ()); ++fii)  // fii = full indices iterator
+        full_indices[fii] = fii;
+
+      // Set up the sorted input indices
+      std::vector<int> sorted_input_indices = *indices_;
+      std::sort (sorted_input_indices.begin (), sorted_input_indices.end ());
+
+      // Store the difference in removed_indices
+      removed_indices_->clear ();
+      set_difference (full_indices.begin (), full_indices.end (), sorted_input_indices.begin (), sorted_input_indices.end (), inserter (*removed_indices_, removed_indices_->begin ()));
```
Comment (taketwo) [inline filters/src/extract_indices.cpp]: I wonder why compilation does not fail without `std::` for `set_difference` and `inserter`. Did we do `using namespace std` somewhere? :confused: 

Anyway, please add `std::`, we try to always qualify standard library usages. (Same below.)

Comment (kunaltyagi) [reply filters/src/extract_indices.cpp]: > I wonder why compilation does not fail without std:: for `set_difference` and `inserter`

That'd be ADL (Argument Dependent Lookup) which looks into the namespace domain of the object if there's no other match found. Hence the trick of `using std::Function; Function(stuff)` which allows a better match named `Function` to still replace (read: specialize) the std implementation

**Comment by kunaltyagi** [comment]
Let's merge this?

**Pull request merged**